File: sensor/src/extraction/dynamic-import.ts
/**
 * Dynamic import() / require() detection and constant-folding for JS/TS.
 *
 * The JS/TS extractors only recognized STATIC `import ... from '...'`
 * (importTypes: ['import_statement']) — dynamic `import(<expr>)` and
 * CommonJS `require(<expr>)` were invisible, producing false-negative
 * `imports` edges (ADR 0048, Lattice sensor correctness fix b). The oracle
 * shape this exists to solve:
 *
 *   import { dirname, join, resolve } from "node:path";
 *   export const ROOT = resolve(import.meta.dirname, "..", "..");
 *   export const CONTROL_LIB = join(ROOT, "lib", "orchestrate", "control-record.mjs");
 *   export const loadControl = () => import(CONTROL_LIB);
 *
 * which requires 3 folding steps: `import.meta.dirname` -> a project-relative
 * dirname, `resolve`/`join` -> path arithmetic, and identifier -> a same-file
 * module-level `const` binding (recursively, since CONTROL_LIB itself refers
 * to ROOT).
 *
 * This module is PURE detection/folding logic — no side effects, no
 * ExtractionContext dependency — so it's usable from (and unit-testable
 * independent of) the core extractor's `extractCall` dispatch, which is the
 * single choke point both AST walkers (the declaration-level `visitNode`
 * ladder AND the function-body-only `visitForCallsAndStructure`) funnel
 * every `call_expression` through. See the call site in tree-sitter.ts for
 * why: a `visitNode`-only hook (the Lua/Ruby `require()` precedent) never
 * fires for a call nested inside a function body — which is exactly where
 * `() => import(CONTROL_LIB)` lives.
 */

import type { Node as SyntaxNode } from 'web-tree-sitter';
import * as posixPath from 'path';
import { getChildByField, getNodeText } from './tree-sitter-helpers';

/**
 * Sentinel `referenceName` for a dynamic import/require whose argument could
 * NOT be statically folded (a variable, a non-`join`/`resolve` function call,
 * a template literal with substitutions, etc.). Deliberately illegal as both
 * a JS identifier and a file basename (angle brackets, colon) so it can NEVER
 * coincidentally equal a real symbol/import-local name — `hasAnyPossibleMatch`
 * and `matchesAnyImport` in the resolver are both name-equality checks, so an
 * unmatchable sentinel guarantees this never resolves via name-matching into
 * an accidental wrong edge. It is pushed as a normal unresolved reference so
 * it goes through the standard resolve-attempt -> mark `status='failed'`
 * lifecycle (ADR 0048's "no silent fallback" requirement) — a `codegraph
 * status`/`unresolved_refs` query surfaces it exactly like any other
 * unresolvable ref, grouped under this one name.
 */
export const DYNAMIC_IMPORT_UNRESOLVED_MARKER = '<dynamic-import:unresolved>';

/** Recursion/cycle guard for identifier -> const-initializer folding. */
const MAX_FOLD_DEPTH = 8;

/** True when `node` is a dynamic `import(...)` call (`function` field is the bare `import` keyword node, not an identifier). */
export function isDynamicImportCall(node: SyntaxNode): boolean {
  if (node.type !== 'call_expression') return false;
  const fn = getChildByField(node, 'function');
  return fn?.type === 'import';
}

/**
 * True when `node` is a bare `require(...)` call. The callee must be a plain
 * `identifier` reading "require" — `x.require(...)` / `foo.require(...)`
 * (a `member_expression` callee) is excluded, matching the Lua precedent's
 * dotted-callee exclusion.
 */
export function isRequireCall(node: SyntaxNode, source: string): boolean {
  if (node.type !== 'call_expression') return false;
  const fn = getChildByField(node, 'function');
  if (!fn || fn.type !== 'identifier') return false;
  return getNodeText(fn, source) === 'require';
}

/**
 * Attempt to fold the single argument of a dynamic import()/require() call
 * to a constant specifier string. Returns null (folding failure, NOT an
 * error) when the argument isn't statically determinable — a bare variable,
 * a function call other than `join`/`resolve`, a template literal with
 * substitutions, multiple/zero arguments, etc.
 *
 * The returned string is used AS-IS as the `imports` unresolved reference's
 * `referenceName` — a relative literal (`./x.mjs`) resolves through the same
 * relative-specifier path static `import` statements use (Strategy 2,
 * resolveViaImport); a project-relative path produced by `join`/`resolve`
 * arithmetic (`lib/orchestrate/control-record.mjs`) instead lands via
 * Strategy 3's `matchByFilePath` exact `filePath` match. Both are legitimate,
 * pre-existing resolution paths — this function doesn't need to know which
 * one a given result will take.
 */
export function foldDynamicImportArg(
  callNode: SyntaxNode,
  source: string,
  filePath: string
): string | null {
  const argsNode = getChildByField(callNode, 'arguments');
  if (!argsNode) return null;
  const args = argsNode.namedChildren.filter((n): n is SyntaxNode => !!n);
  // A dynamic import/require legitimately takes exactly one argument. Zero
  // or multiple (e.g. a webpack magic-comment second arg, which parses as a
  // comment, not a real 2nd arg, so this is effectively "not exactly one
  // real argument") is outside what we can safely fold — bail rather than
  // guess which argument matters.
  if (args.length !== 1) return null;
  return foldExpr(args[0]!, source, filePath, 0, new Set());
}

function foldExpr(
  node: SyntaxNode,
  source: string,
  filePath: string,
  depth: number,
  visitedIdentifiers: ReadonlySet<string>
): string | null {
  if (depth > MAX_FOLD_DEPTH) return null;

  switch (node.type) {
    case 'string': {
      // `string > string_fragment` gives the bare content, sidestepping
      // quote-char stripping/escaping edge cases (mirrors lua.ts's
      // requireModule, which reads string_content the same way).
      const fragment = node.namedChildren.find((c) => c?.type === 'string_fragment');
      return fragment ? getNodeText(fragment, source) : '';
    }

    case 'template_string': {
      // Fold ONLY when every named child is a plain string_fragment — a
      // `template_substitution` (`${expr}`) OR any other child (e.g. an
      // `escape_sequence`) makes the literal's runtime value unknowable
      // from source text alone, so bail rather than approximate.
      const parts: string[] = [];
      for (const child of node.namedChildren) {
        if (!child) continue;
        if (child.type !== 'string_fragment') return null;
        parts.push(getNodeText(child, source));
      }
      return parts.join('');
    }

    case 'identifier': {
      const name = getNodeText(node, source);
      if (name === '__dirname') return projectRelativeDirname(filePath);
      if (visitedIdentifiers.has(name)) return null; // cycle guard (A = B, B = A)
      const initializer = findModuleLevelConstInitializer(node, name, source);
      if (!initializer) return null;
      return foldExpr(initializer, source, filePath, depth + 1, new Set([...visitedIdentifiers, name]));
    }

    case 'member_expression': {
      // `import.meta.dirname` — the ONLY member_expression shape this folds.
      // tree-sitter-javascript parses `import.meta` as a single `meta_property`
      // node (not a nested member_expression), so this is a one-level check.
      const object = getChildByField(node, 'object');
      const property = getChildByField(node, 'property');
      if (
        object?.type === 'meta_property' &&
        getNodeText(object, source) === 'import.meta' &&
        property &&
        getNodeText(property, source) === 'dirname'
      ) {
        return projectRelativeDirname(filePath);
      }
      return null;
    }

    case 'call_expression': {
      // Bare `join(...)`/`resolve(...)` OR a member form (`path.join(...)`,
      // `nodePath.resolve(...)`) — the receiver identifier is intentionally
      // unchecked (spec: "裸 or path.join / path.resolve member 形"), only
      // the invoked method name matters.
      const fn = getChildByField(node, 'function');
      const argsNode = getChildByField(node, 'arguments');
      if (!fn || !argsNode) return null;
      let methodName: string | null = null;
      if (fn.type === 'identifier') {
        methodName = getNodeText(fn, source);
      } else if (fn.type === 'member_expression') {
        const property = getChildByField(fn, 'property');
        methodName = property ? getNodeText(property, source) : null;
      }
      if (methodName !== 'join' && methodName !== 'resolve') return null;

      const parts: string[] = [];
      for (const arg of argsNode.namedChildren) {
        if (!arg) continue;
        const folded = foldExpr(arg, source, filePath, depth + 1, visitedIdentifiers);
        if (folded === null) return null;
        parts.push(folded);
      }
      return foldJoinResolve(parts);
    }

    case 'parenthesized_expression': {
      // `(CONTROL_LIB)` — unwrap defensively; harmless if the grammar never
      // actually nests one here for our shapes.
      const inner = node.namedChildren.find((c) => !!c);
      return inner ? foldExpr(inner, source, filePath, depth + 1, visitedIdentifiers) : null;
    }

    default:
      // Function calls other than join/resolve, binary expressions
      // (`'./y' + 'z'`), member access on anything else, etc. — not
      // statically foldable. Folding failure, not an error.
      return null;
  }
}

/**
 * Find a module-level (top-of-file) `const NAME = <expr>` declarator's
 * initializer, honouring an `export` wrapper. Deliberately scoped to
 * Program's direct children only (not nested inside functions/blocks) —
 * matches the spec's "module-level const" requirement. `let`/`var` are
 * excluded (spec: "再代入されうる let/var は畳まない") since they can be
 * reassigned and the declaration-site value is not necessarily the
 * call-site value.
 */
function findModuleLevelConstInitializer(
  anyNodeInTree: SyntaxNode,
  name: string,
  source: string
): SyntaxNode | null {
  const root = anyNodeInTree.tree.rootNode;
  for (let i = 0; i < root.namedChildCount; i++) {
    let decl = root.namedChild(i);
    if (!decl) continue;
    if (decl.type === 'export_statement') {
      const inner = getChildByField(decl, 'declaration');
      if (inner) decl = inner;
    }
    if (decl.type !== 'lexical_declaration') continue;
    const kindNode = getChildByField(decl, 'kind');
    if (kindNode?.type !== 'const') continue;

    for (let j = 0; j < decl.namedChildCount; j++) {
      const declarator = decl.namedChild(j);
      if (declarator?.type !== 'variable_declarator') continue;
      const nameNode = getChildByField(declarator, 'name');
      if (nameNode && getNodeText(nameNode, source) === name) {
        return getChildByField(declarator, 'value');
      }
    }
  }
  return null;
}

/** Project-relative dirname of `filePath` (posix, root as `''` not `'.'`). */
function projectRelativeDirname(filePath: string): string {
  const dir = posixPath.posix.dirname(filePath.replace(/\\/g, '/'));
  return dir === '.' ? '' : dir;
}

/**
 * Fold `join`/`resolve` arguments into one project-relative path. Both are
 * treated identically — we operate purely in project-relative "pseudo-
 * absolute" path space (there's no real filesystem CWD at extraction time),
 * so `resolve`'s "absolute result" semantics and `join`'s "relative to first
 * arg" semantics collapse to the same segment-join-and-normalize arithmetic.
 * `..` is resolved (per spec); a result that escapes the project root is a
 * folding failure — we can't represent "outside the project" as a valid
 * project-relative file path, and guessing would risk a wrong edge.
 */
function foldJoinResolve(parts: string[]): string | null {
  let result = parts[0] ?? '';
  for (let i = 1; i < parts.length; i++) {
    const part = parts[i]!;
    // A real filesystem-absolute segment (rare in this idiom, and not
    // something we can map into project-relative space) — bail.
    if (part.startsWith('/')) return null;
    result = result ? `${result}/${part}` : part;
  }
  const normalized = posixPath.posix.normalize(result);
  if (normalized === '..' || normalized.startsWith('../')) return null;
  return normalized === '.' ? '' : normalized;
}

File: sensor/lattice-sensor-kernel/src/tsjs/dynamic_import.rs
//! Dynamic `import(...)` / CommonJS `require(...)` specifier folding — a
//! faithful Rust port of src/extraction/dynamic-import.ts (ADR 0048).
//!
//! Folding failure is NOT an error: an argument that isn't statically
//! determinable returns None, and the caller records a visible unresolved
//! sentinel rather than fabricating an edge.

use tree_sitter::Node;

/// Sentinel `referenceName` for an unfoldable specifier. Chosen so it can
/// never accidentally name-match a real symbol or import.
pub const DYNAMIC_IMPORT_UNRESOLVED_MARKER: &str = "<dynamic-import:unresolved>";

/// Recursion/cycle guard for identifier -> const-initializer folding.
const MAX_FOLD_DEPTH: u32 = 8;

fn text<'a>(node: Node, src: &'a str) -> &'a str {
    &src[node.byte_range()]
}

/// True when `node` is a dynamic `import(...)` call — the `function` field is
/// the bare `import` keyword node, not an identifier.
pub fn is_dynamic_import_call(node: Node) -> bool {
    node.kind() == "call_expression"
        && node.child_by_field_name("function").map(|f| f.kind()) == Some("import")
}

/// True when `node` is a bare `require(...)` call. A dotted callee
/// (`x.require(...)`) is excluded, matching the Lua precedent.
pub fn is_require_call(node: Node, src: &str) -> bool {
    if node.kind() != "call_expression" {
        return false;
    }
    let Some(f) = node.child_by_field_name("function") else { return false };
    f.kind() == "identifier" && text(f, src) == "require"
}

/// Fold the single argument of a dynamic import()/require() call to a constant
/// specifier. Zero or multiple arguments is outside what can safely be folded.
pub fn fold_dynamic_import_arg(call: Node, src: &str, file_path: &str) -> Option<String> {
    let args_node = call.child_by_field_name("arguments")?;
    let args: Vec<Node> = (0..args_node.named_child_count())
        .filter_map(|i| args_node.named_child(i))
        .collect();
    if args.len() != 1 {
        return None;
    }
    fold_expr(args[0], src, file_path, 0, &[])
}

/// Same folding for one already-selected argument, without the argument-count
/// gate — `fork(mod, args, opts)` legitimately takes more than one.
pub fn fold_constant_expr(node: Node, src: &str, file_path: &str) -> Option<String> {
    fold_expr(node, src, file_path, 0, &[])
}

fn fold_expr(
    node: Node,
    src: &str,
    file_path: &str,
    depth: u32,
    visited: &[String],
) -> Option<String> {
    if depth > MAX_FOLD_DEPTH {
        return None;
    }
    match node.kind() {
        // `string > string_fragment` gives the bare content, sidestepping
        // quote-char stripping and escape edge cases.
        "string" => {
            let fragment = (0..node.named_child_count())
                .filter_map(|i| node.named_child(i))
                .find(|c| c.kind() == "string_fragment");
            Some(fragment.map(|f| text(f, src).to_string()).unwrap_or_default())
        }

        // Fold ONLY when every named child is a plain string_fragment — a
        // `${expr}` substitution (or an escape_sequence) makes the runtime
        // value unknowable from source text alone.
        "template_string" => {
            let mut parts = String::new();
            for i in 0..node.named_child_count() {
                let Some(child) = node.named_child(i) else { continue };
                if child.kind() != "string_fragment" {
                    return None;
                }
                parts.push_str(text(child, src));
            }
            Some(parts)
        }

        "identifier" => {
            let name = text(node, src);
            if name == "__dirname" {
                return Some(project_relative_dirname(file_path));
            }
            if visited.iter().any(|v| v == name) {
                return None; // cycle guard (A = B, B = A)
            }
            let initializer = find_module_level_const_initializer(node, name, src)?;
            let mut next: Vec<String> = visited.to_vec();
            next.push(name.to_string());
            fold_expr(initializer, src, file_path, depth + 1, &next)
        }

        // `import.meta.dirname` — the ONLY member_expression shape this folds.
        // tree-sitter parses `import.meta` as a single `meta_property` node.
        "member_expression" => {
            let object = node.child_by_field_name("object")?;
            let property = node.child_by_field_name("property")?;
            if object.kind() == "meta_property"
                && text(object, src) == "import.meta"
                && text(property, src) == "dirname"
            {
                return Some(project_relative_dirname(file_path));
            }
            None
        }

        // Bare `join(...)`/`resolve(...)` or a member form (`path.join(...)`).
        // The receiver is intentionally unchecked — only the method name matters.
        "call_expression" => {
            let f = node.child_by_field_name("function")?;
            let args_node = node.child_by_field_name("arguments")?;
            let method = match f.kind() {
                "identifier" => Some(text(f, src).to_string()),
                "member_expression" => f
                    .child_by_field_name("property")
                    .map(|p| text(p, src).to_string()),
                _ => None,
            }?;
            if method != "join" && method != "resolve" {
                return None;
            }
            let mut parts: Vec<String> = Vec::new();
            for i in 0..args_node.named_child_count() {
                let Some(arg) = args_node.named_child(i) else { continue };
                parts.push(fold_expr(arg, src, file_path, depth + 1, visited)?);
            }
            fold_join_resolve(&parts)
        }

        // `(CONTROL_LIB)` — unwrap defensively.
        "parenthesized_expression" => {
            let inner = (0..node.named_child_count()).find_map(|i| node.named_child(i))?;
            fold_expr(inner, src, file_path, depth + 1, visited)
        }

        // Other calls, binary expressions (`'./y' + 'z'`), member access on
        // anything else — not statically foldable. Failure, not an error.
        _ => None,
    }
}

/// Module-level `const NAME = <expr>` initializer, honouring an `export`
/// wrapper. Program's direct children only — `let`/`var` are excluded since
/// they can be reassigned and the declaration-site value is not necessarily
/// the call-site value.
fn find_module_level_const_initializer<'t>(
    any_node: Node<'t>,
    name: &str,
    src: &str,
) -> Option<Node<'t>> {
    // The Rust binding has no Node::tree(); walk up to the root instead.
    let mut root = any_node;
    while let Some(parent) = root.parent() {
        root = parent;
    }
    for i in 0..root.named_child_count() {
        let Some(mut decl) = root.named_child(i) else { continue };
        if decl.kind() == "export_statement" {
            if let Some(inner) = decl.child_by_field_name("declaration") {
                decl = inner;
            }
        }
        if decl.kind() != "lexical_declaration" {
            continue;
        }
        if decl.child_by_field_name("kind").map(|k| k.kind()) != Some("const") {
            continue;
        }
        for j in 0..decl.named_child_count() {
            let Some(declarator) = decl.named_child(j) else { continue };
            if declarator.kind() != "variable_declarator" {
                continue;
            }
            let Some(name_node) = declarator.child_by_field_name("name") else { continue };
            if text(name_node, src) == name {
                return declarator.child_by_field_name("value");
            }
        }
    }
    None
}

/// Project-relative dirname (posix, root as `""` not `"."`).
fn project_relative_dirname(file_path: &str) -> String {
    let normalized = file_path.replace('\\', "/");
    match normalized.rfind('/') {
        // posix.dirname("/a") is "/", and "a/" behaves as its own parent —
        // both collapse to the leading segment the same way node's does.
        Some(0) => "/".to_string(),
        Some(i) => normalized[..i].to_string(),
        None => String::new(), // dirname → "." → ""
    }
}

/// Fold `join`/`resolve` arguments into one project-relative path. Both are
/// treated identically: there is no filesystem CWD at extraction time, so
/// `resolve`'s absolute semantics and `join`'s relative semantics collapse to
/// the same segment arithmetic. A result escaping the project root is a
/// folding failure — guessing would risk a wrong edge.
fn fold_join_resolve(parts: &[String]) -> Option<String> {
    let mut result = parts.first().cloned().unwrap_or_default();
    for part in parts.iter().skip(1) {
        // A real filesystem-absolute segment can't be mapped into
        // project-relative space.
        if part.starts_with('/') {
            return None;
        }
        result = if result.is_empty() { part.clone() } else { format!("{result}/{part}") };
    }
    let normalized = posix_normalize(&result);
    if normalized == ".." || normalized.starts_with("../") {
        return None;
    }
    Some(if normalized == "." { String::new() } else { normalized })
}

/// `path.posix.normalize` for the subset this folding produces: resolves `.`
/// and `..`, collapses repeated separators, preserves a leading `/` and a
/// leading `..` run, and returns `.` for an empty relative result.
pub(super) fn posix_normalize_for_spawn(input: &str) -> String {
    posix_normalize(input)
}

fn posix_normalize(input: &str) -> String {
    let absolute = input.starts_with('/');
    let trailing_slash = input.len() > 1 && input.ends_with('/');
    let mut out: Vec<&str> = Vec::new();
    for segment in input.split('/') {
        match segment {
            "" | "." => {}
            ".." => {
                match out.last() {
                    Some(&last) if last != ".." => {
                        out.pop();
                    }
                    _ => {
                        // Leading `..` survives only in relative paths; above
                        // the root of an absolute path it is dropped.
                        if !absolute {
                            out.push("..");
                        }
                    }
                }
            }
            other => out.push(other),
        }
    }
    let mut joined = out.join("/");
    if joined.is_empty() {
        return if absolute { "/".to_string() } else { ".".to_string() };
    }
    if trailing_slash {
        joined.push('/');
    }
    if absolute {
        format!("/{joined}")
    } else {
        joined
    }
}
